lib/Feature/FeatureModelParser.cpp

```diff
@@ -225,4 +252,361 @@ std::unique_ptr<xmlDoc, void (*)(xmlDocPtr)> FeatureModelXmlParser::parseDoc() {
+        continue;
+      }
+
+      // For every line, count the indentation
+      // not more than 1 additional indentations are allowed to the original one
+      // However, we may have arbitrarily less indentations
+      if ((To.find(':', 0 )) == std::string::npos) {
+        std::cerr << "Colon is missing in line" << To << std::endl;
+        return false;
+      }
+
+      int CurrentIndentationLevel = countOccurrences(readUntil(To, ':', 0), Indentation);
+      int Diff = CurrentIndentationLevel - LastIndentationLevel;
+
+      // Remember the root indentation for later checks
+      if (LastIndentationLevel == -1) {
+        RootIndentation = CurrentIndentationLevel;
+      }
+
+      if ((LastIndentationLevel != -1) && Diff > 1) {
+        std::cerr << "Indentation error in feature tree in line " << To << std::endl;
+        return false;
+      }
+
+      // Move pointer to first character after indentation
+      // The first character has to be a colon followed by the type of
+      // the feature (m for mandatory, o for optional, a for alternative)
+      std::string::size_type Pos = CurrentIndentationLevel * Indentation.length() + 2;
+      std::optional<std::tuple<int, int>> Cardinalities;
+
+      switch (To.at(Pos - 1)) {
+      case 'o':
+        // Code for optional
+        Opt = true;
+        break;
+      case 'g':
+        // Code for an or group with different cardinalities
+        Opt = false;
+        // Extract the cardinality
+        Cardinalities = extractCardinality(To);
+        if (!Cardinalities.has_value()) {
+          return false;
+        }
+        break;
+      case ' ':
+        // Code for alternative child
+        Pos--;
+        break;
+      }
+      // Extract the name
+      Name = readUntil(To, ' ', Pos + 1);
+
+      // Note that we ignore the ID and use the name of the feature
+      // as unique identifier.
+      if (Name.find_first_of('(') != std::string::npos) {
+        Name = readUntil(Name, '(', 0);
+      }
+
+      // Remove the cardinality
+      if (Name.find_first_of('[') != std::string::npos) {
+        Name = readUntil(Name, '[', 0);
+      }
+
+      // If there is no name, provide an artificial one
+      if (Name.empty()) {
+        OrGroupCounter++;
+        Name = "group_" + std::to_string(OrGroupCounter);
+      }
+
+      // Create the feature
+      FMB.makeFeature<BinaryFeature>(Name, Opt);
+      IndentationToParentMapping[CurrentIndentationLevel] = Name;
+
+      // Add parent from the upper indentation level if there is one
+      if (LastIndentationLevel != -1 && CurrentIndentationLevel == RootIndentation) {
+        std::cerr << "Only one feature can be root and have the same indentation as root." << std::endl;
+        return false;
+      }
+
+      if (LastIndentationLevel != -1) {
+        auto Parent = IndentationToParentMapping.find(CurrentIndentationLevel - 1);
+        assert(Parent != IndentationToParentMapping.end());
```

Comment:
Hmm, is this more the case that "this should normally never happen" or "a wrong formatted file causes this case". In case of the second one, I would suggest different error handling because if this case happens in an optimized build -- no asserts -- the dereference bellow is UB and could/will crashe hard.